text/0000-no-no-std-standard.md

```diff
@@ -0,0 +1,100 @@
+
+As of this writing, there are [28 crates] on crates.io (0.4% of 7,426 total) with a feature that contains `no_` as a substring. Out of these, 21 of them have `no_std`. As for the remaining seven:
+
+- `gif` has a `raii_no_panic` feature. When enabled, any I/O errors are ignored when the write buffer is dropped. In the author's opinion, the library should expose an explicit `.flush()` method and avoid panicking altogether. Verdict: **change**.
+
+- `lazy_static` has a `spin_no_std` feature which replaces the usual implementation with a spin lock. This has a similar use case to `no_std`, and so has the same verdict of **change**.
+
+- `libbindgen` has a `assert_no_dangling_items` feature. This appears to be a [debug tool][libbindgen commit]. Verdict: **keep**.
+
+- `linked-tail-list` has a `test_no_validate` feature, which disables some sanity checks. In the author's opinion, this library should either make this positive (like `libbindgen`) or use `#[cfg(test)]` instead. Verdict: **change**.
+
+- `quickercheck` has a feature called `no_function_casts` which changes some impl items to use the `Fn` trait directly. Given how it's used, the feature could be renamed to `nightly` instead. There's no clear benefit to this though. Verdict: **keep**.
+
+- `ralloc` has two features, `unsafe_no_mutex_lock` and `no_log_lock`, which trade thread safety for increased performance. But even if one crate knows that it uses `ralloc` in a single-threaded way, it cannot guarantee that other crates do as well. On the other hand, given the syntax:
+
+    ```toml
+    [dependencies.ralloc]
+    default-features = false
+    features = []  # disable: unsafe_mutex_lock, log_lock
+    ```
+
+    it feels strange opting out of safety by *omitting* a word.
+
+    Verdict: **uncertain**.
+
+- `rscam` has a feature called `no_wrapper` which disables linking to the C library. This should be changed to a positive `use_libv4l2` feature instead. **Change**.
+
+Given the variation between these crates, we cannot justify a blanket ban on all features that contain the string `no_`. We conclude that the best solution is to blacklist `no_std` only.
+
+[28 crates]: https://gist.github.com/lfairy/5767bd29de07554e059981e18449ba44
+[libbindgen commit]: https://github.com/servo/rust-bindgen/commit/15c687e0f1bec97448168f981ec93d207884a775
+
+# Detailed design
+[design]: #detailed-design
+
+This change should be rolled out incrementally, in a similar way to the [ban on wildcard dependencies](./1241-no-wildcard-deps.md).
+
+In the next stable Rust release, Cargo will issue warnings for all crates that have a `no_std` feature when publishing. However, the publish will still succeed.
+
+In time with the stable release after that, crates.io will be updated to reject crates which declare this feature. Note that the check will only run on the crates.io server: other package hosts will be free to implement another policy.
```

Comment:
What happens to the crates and versions that have already been published when this change is rolled out?

What happens to people running old versions of Rust and Cargo, even from before the version of cargo that implements the warnings, if they have a `no_std` feature and publish to crates.io?

I don't think we've done a deploy of crates.io that has to coincide with a Rust release before.... :-/ I can imagine possibilities where this might be necessary, but it makes me nervous....